modules/apps/remote-app/remote-app-web/src/main/java/com/liferay/remote/app/web/internal/servlet/taglib/RemoteAppTopHeadDynamicInclude.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.remote.app.web.internal.servlet.taglib;
+
+import com.liferay.portal.kernel.model.Portlet;
+import com.liferay.portal.kernel.servlet.taglib.BaseDynamicInclude;
+import com.liferay.portal.kernel.servlet.taglib.DynamicInclude;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(
+	immediate = true,
+	service = {DynamicInclude.class, RemoteAppTopHeadDynamicInclude.class}
+)
+public class RemoteAppTopHeadDynamicInclude extends BaseDynamicInclude {
+
+	@Override
+	public void include(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, String key)
+		throws IOException {
+
+		PrintWriter printWriter = httpServletResponse.getWriter();
+
+		for (String url : _urls.keySet()) {
+			printWriter.println(
+				"<script type=\"module\" src=\"" + url + "\"></script>");
+		}
+	}
+
+	@Override
+	public void register(DynamicIncludeRegistry dynamicIncludeRegistry) {
+		dynamicIncludeRegistry.register("/html/common/themes/top_head.jsp#pre");
+	}
+
+	public synchronized void registerURLs(String portletName, String[] urls) {
+		synchronized (this) {
+			_portletURLsMap.put(portletName, urls);
+
+			for (String url : urls) {
+				Integer referenceCount = _urls.get(url);
```

Comment:
Thinking if it can be simplify using map.getOrDefault or map.putIfAbsent but is good anyway

Reply:
Good catch. I'll try it :+1:

Reply:
I think it's not possible without using the synchronized block because we need to remove the reference counts from the concurrent map when they reach to zero but that's not possible using atomic operations because AtomicInteger does not implement equals, so there's no atomic way to remove it from the concurrent map when it is zero.

I'm not saying there isn't a way to do it, but it's not obvious and I don't think those `synchronized` will create any bottleneck given how small the exclusion blocks are and how they will be executed:

- We will only have bursts of calls to `registerURLs` from single threads (when the portal starts, for example).
- Regarding `removedService` it will only be called massively when the portal shuts down